Skyline: Improve failure reporting in AssertEx.FileEquals

brendanx67 opened the pull request:
This greatly improves failure reporting for AssertEx.FileEquals() where the old code used to just report the two lines where the comparison failed, the new code gives a lot more detail on the internal reasons when paths and column tolerances are involved. This code also adds a class ColumnTolerances to formalize the tolerance evaluation and introduces a tolerance that applies to the mantissa of the number in scientific notation to better capture limitations in precision and potential rounding error.

Before (mistakenly interpreted as a path difference):

Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.Fail failed.  Diff found at line 225 position 0: expected
Orbi3_SA_IP_pHis3_01.mzML  17836  17876  8  3  3495.7076  1166.5775  1.050319E+08  2.961274E+08  110.6513  110.8777  110.7155  0.9989  _  17848  H254C156N43O46S1[-2.150306]  mass3495.7076_RT110.7155
actual
z:\download\Perftests\Label-free\converted\Orbi3_SA_IP_pHis3_01.mzML  17836  17876  8  3  3495.7076  1166.5775  1.050319E+08  2.961275E+08  110.6513  110.8777  110.7155  0.9989  _  17848  H254C156N43O46S1[-2.150306]  mass3495.7076_RT110.7155

After (skipping thousands of lines that match by the specified precision and more clearly showing the problem):

Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.Fail failed.  Diff found at line 7723 position 5:
expected
Orbi3_SA_IP_pHis3_01.mzML  2350  2371  5  1  863.4852  864.4928  1732674  5650320  15.8053  15.9211  15.8631  0.9988  _  2361  H86C38N10O11[+4.837413]  mass863.4852_RT15.8631
actual
z:\download\Perftests\Label-free\converted\Orbi3_SA_IP_pHis3_01.mzML  8456  8477  6  2  1228.6413  615.3279  1732675  8681932  51.0093  51.1453  51.1453  0.9942  _  8477  H102C55N15O16[-0.121619]  mass1228.6413_RT51.1453
expected with paths removed
path  2350  2371  5  1  863.4852  864.4928  1732674  5650320  15.8053  15.9211  15.8631  0.9988  _  2361  H86C38N10O11[+4.837413]  mass863.4852_RT15.8631
actual with paths removed
path  8456  8477  6  2  1228.6413  615.3279  1732675  8681932  51.0093  51.1453  51.1453  0.9942  _  8477  H102C55N15O16[-0.121619]  mass1228.6413_RT51.1453
matching prefix: 'path  '
matching suffix: ''
decimal matching: Expected decimal mantissa: 2.35 does not match actual 8.456 to within 0.00015015

bspratt commented:
Looks good, other than I really do not think we should be using the scientific notation logic on integers. It took me a little while to understand that it was complaining about "2350" vs "8456". When I think about it, I'd argue for using the E logic only if one or the other text representations does in fact use scientific notation.

And it seems like we could be reporting the column index when comparing TSVs, but that's a nice-to-have.

brendanx67 commented:
I think the integer issue can still be a precision thing. I did realize
that I could be reporting the original text and the delta multiplied by the
exponent which would be clearer in the integer case. Maybe I will do that
before merging. Just wanted to get this code in a PR, because I felt it
helped immensely to understand what was going on, and in the future avoid
thinking about paths when they were not the issue.

On Wed, Jun 19, 2024 at 2:13 PM Brian Pratt ***@***.***>
wrote:

> Looks good, other than I really do not think we should be using the
> scientific notation logic on integers. It took me a little while to
> understand that it was complaining about "2350" vs "8456". When I think
> about it, I'd argue for using the E logic only if one or the other text
> representations does in fact use scientific notation.
>
> And it seems like we could be reporting the column index when comparing
> TSVs, but that's a nice-to-have.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ProteoWizard/pwiz/pull/3021#issuecomment-2179462757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYBWUCHBOJEJKUMU6LRZEDZIHYARAVCNFSM6AAAAABJSVHKGCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNZZGQ3DENZVG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

brendanx67 commented:
Merged master and addressed the `ForceScientificNotation` review feedback. Here's what the branch now contains relative to master:

**Improved `AssertEx.FileEquals` failure reporting:**
- Reports line number and character position of first difference
- Shows both expected and actual lines with matching prefix/suffix for easy visual diffing
- When path differences are removed, shows both original and cleaned-up text
- When numeric tolerance comparison fails, includes the specific decimal mismatch detail

**Per-column tolerance support via typed classes:**
- New `ColumnTolerances` and `ColumnToleranceValue` classes replace the raw `Dictionary<int, double>` used by `CommonTextUtil.LinesEquivalentIgnoringTimeStampsAndGUIDs`
- Cleaner call-site API: `new AssertEx.ColumnTolerances(0.00015)` with optional `AddTolerance(column, value)` for per-column overrides

**`ForceScientificNotation` removed** (per Brian's review):
- No longer converts integers like "2350" to "2.350000E+003" for comparison
- Mantissa/exponent comparison only activates when both values naturally contain scientific notation ('E'/'e')
- Plain numbers are compared as doubles with tolerance — simpler and sufficient for our needs

**Picked up master's timestamp regex fix:**
- Timezone portion is now optional in ISO timestamp matching, so bare timestamps like `2020-07-10T10:40:03` are recognized

Build succeeds with zero warnings. `CodeInspection` and `ChromatogramExporterTest` pass.

brendanx67 commented on pwiz_tools/Skyline/TestPerf/FeatureDetectionTest.cs:
```diff
@@ -452,7 +452,7 @@ void ExpectError(Action act)
                 var hkActualFilePath = hkExpectedFilePath.Replace(expectedHardklorFiles, Path.Combine(expectedHardklorFiles, @".."));
-                var columnTolerances = new Dictionary<int, double>() { { -1, .00015 } }; // Allow a little rounding wiggle in the decimal values
+                var columnTolerances = new AssertEx.ColumnTolerances(0.00015); // Allow a little rounding wiggle in the numeric values
                 AssertEx.FileEquals(hkExpectedFilePath, hkActualFilePath, columnTolerances, true);
```
Moved outside loop

brendanx67 commented on pwiz_tools/Skyline/TestUtil/AssertEx.cs:
```diff
@@ -961,6 +1002,156 @@ private static void RemovePathDifferences(ref string lineExpected, ref string li
+        {
+            private ColumnToleranceValue _defaultTolerance { get; }
+            private Dictionary<int, ColumnToleranceValue> _explicitTolerances = new Dictionary<int, ColumnToleranceValue>();
+
```
Good catch. Turned both values into private readonly instance variables.

brendanx67 commented on pwiz_tools/Skyline/TestUtil/AssertEx.cs:
```diff
@@ -961,6 +1002,156 @@ private static void RemovePathDifferences(ref string lineExpected, ref string li
+            if (matchExpected.Success && matchActual.Success
+                                      && Equals(matchExpected.Groups[1].ToString(), matchActual.Groups[1].ToString())
+                                      && Equals(matchExpected.Groups[2].ToString(), matchActual.Groups[2].ToString()))
+            {
```
The duplication is intentional — the local copy adds the \ parameter that \ doesn't support. Changing \'s signature would be a shared library change affecting other consumers. The GUID/timestamp regexes are stable and unlikely to drift.

brendanx67 commented on pwiz_tools/Skyline/TestUtil/AssertEx.cs:
```diff
@@ -944,13 +984,14 @@ private static void RemovePathDifferences(ref string lineExpected, ref string li
+                            // So, replace the matching paths with visible matching text.
+                            const string pathSubstitutionText = "path";
+                            lineExpected = lineExpected.Replace(pathE, pathSubstitutionText);
+                            lineActual = lineActual.Replace(pathA, pathSubstitutionText);
```
This is pre-existing code not changed by this PR. The current behavior is intentional — it replaces the full column value to produce a clean diff display. Changing to replace only the path substring would be a behavior change that could affect existing tests.
